src/fsharp/IlxGen.fs

```diff
@@ -834,7 +841,10 @@ and IlxGenEnv =
       /// All values in scope
       valsInScope: ValMap<Lazy<ValStorage>>
 
-      /// For optimizing direct tail recursion to a loop - mark says where to branch to. Length is 0 or 1.
+      /// All witnesses in scope
+      witnessesInScope: ImmutableDictionary<TraitWitnessInfo, ValStorage>
```

Comment:
While I'm ok with using `ImmutableDictionary`, we tend to be using F#'s `Map`.

What was the reason for choosing `ImmutableDictionary` instead of `Map` here? Performance?

Reply:
It is because the dictionary gets an IEqualityComparer that captures `TcGlobals`, here
```fsharp
      witnessesInScope= 
          ImmutableDictionary.Create(
               { new IEqualityComparer<_> with 
                      member __.Equals(a, b) = traitKeysAEquiv g TypeEquivEnv.Empty a b
                      member __.GetHashCode(a) = hash a.MemberName
               })
```

If the `TraitWitnessInfo` objects captured `TcGlobals` themselves and could thus implement their own internalized equality/hash logic. However even then we couldn't use a `Map` becuase they are built on sorted keys using IComparer, rather than hashing using IEqualityComparer.

I'll add a note to the source